### ISSUE TITLE
Keyboard shortcuts for zoom in/zoom broken on some keyboard layouts

### ISSUE THREAD

**Issue by bagage** [openstreetmap/iD#2885]
It seems related to #2389 but zoom in/out shortcuts are NOT `+`/`-` keys but instead layout specific:
1. on qwerty layout, `+` is actually `=` (keycode 21), and `-` is `-`  (keycode 20)
2. on azerty layout, `+` is `=` (keycode 21) and `-` is `^` (keycode 15)
3. on my custom layout (a bépo French layout variant), `+` is `=` (keycode 16) and `-` is `^` (keycode 17).

Please note that this is specific to Firefox (version 41.0.1 here) though, since Epiphany is working as expected with all layouts.

edit: by the way, zooming in and out using shortcuts is not based on the center of the map - should I enter another bug for that?

**Comment by bagage** [comment]
Looks also related to #2812.

**Comment by bhousel** [comment]
Yes this is related to #2812, but I'm ok with having this issue open for the zoom shortcuts, which I think we can possibly fix in the short term as long as we can bind to codes that other layouts send.

**Comment by bhousel** [comment]
> by the way, zooming in and out using shortcuts is not based on the center of the map - should I enter another bug for that?

Weird, I do not see this in any of the browsers that I test.  Where do you see the zoom centered on?

**Comment by bagage** [comment]
This is not specific to any browser, but there is a strange movement when zooming in and out... hard to explain, here is a short recording:
![zoomin](https://cloud.githubusercontent.com/assets/1451988/11996507/af6955dc-aa64-11e5-8203-78994f97f924.gif)

**Comment by bhousel** [comment]
>  there is a strange movement when zooming in and out...

ah, yeah the video makes it clear, and I do see that too.  It's related to [this hacky code](https://github.com/openstreetmap/iD/blob/master/js/id/renderer/map.js#L262-L266) to get D3 to interpolate the zoom.. We can definitely make it better.

**Comment by bagage** [comment]
Oh indeed! That's not a big deal though ;-).

**Comment by bhousel** [comment]
@bagage : Can you test these issues again with the latest http://preview.ideditor.com/master ?  When you test, let me know what browser and os you are using too, because that might matter.

I just merged #3572 (thanks @tyrasd!) which gives us a more reliable way to know which key was pressed in modern browsers.

**Comment by bagage** [comment]
@bhousel Excellent! I did a few tests with **same results across Firefox 52.0.2, [Eolie git master](https://github.com/gnumdk/eolie/issues) and Chromium 58.0.3029.96 on Debian unstable:

Firefox + azerty: can zoom in/out with `+` (21) key AND `-` (15) keys but **also zoom in with `_` (17)**.
Firefox + custom bépo: can zoom in/out with `+` (16) AND `-` (17) keys but **also zoom in with `=` (20)**.
Firefox + qwerty: can zoom with `+` (20) and `-` (21) keys.

Basically it's now working good even if there are some extra shortcuts. Thanks for the good work @bhousel and @tyrasd :).

**Comment by bhousel** [comment]
Good to hear! closing..

The extra shortcuts `_` and `=` are expected because on US keyboard they are shared with `+` and `-`, and I wanted users to be able to use the keys either shifted or not.

<img width="268" alt="screenshot 2017-06-06 11 53 02" src="https://user-images.githubusercontent.com/38784/26838707-e9d27442-4aae-11e7-97ba-83fa49da25bf.png">

**Comment by bagage** [comment]
Yeah that's what I supposed too. Cool!
